libs/evm2wasm/evm2wasm.cpp

```diff
@@ -1,61 +1,761 @@
+        {
+            segment << "(set_global $sp(i32.add(get_global $sp)(i32.const {stackDeta})))\n"_format(
+                "stackDeta"_a = stackDeta * 32);
+        }
+
+        // creates a stack trace
+        if (stackTrace)
+        {
+            segment << "(call $stackTrace(i32.const {pc})(i32.const {opint})( \
+                        i32.const {gasCount})(get_global $sp))\n"_format(
+                "pc"_a = pc, "opint"_a = opint, "gasCount"_a = gasCount);
+        }
+
+        // adds the logic to save the stack pointer before exiting to wiat to for a callback
+        // note, this must be done before the sp is updated above^
+        if (useAsyncAPI && callbackFuncs.find(op.name) != std::end(callbackFuncs))
+        {
+            segment << "(set_global $cb_dest (i32.const {jumpSegmentsLength})) \
+                            (br $done))"_format("jumpSegmentsLength"_a = jumpSegments.size() + 1);
+            jumpSegments.push_back({number : 0, type : "cb_dest"});
+        }
+    }
+
+    endSegment();
+
+    std::string wastStr = wast.str();
+    wast.clear();
+    wast << assembleSegments(jumpSegments) << wastStr << "))";
+
+    auto wastFiles = wastSyncInterface;  // default to synchronous interface
+    if (useAsyncAPI)
+    {
+        wastFiles = wastAsyncInterface;
+    }
+
+    std::vector<std::string> imports;
+    std::vector<std::string> funcs;
+    // inline EVM opcode implemention
+    if (inlineOps)
+    {
+        std::tie(funcs, imports) = resolveFunctions(opcodesUsed, wastFiles);
+    }
+
+    // import stack trace function
+    if (stackTrace)
+    {
+        imports.push_back("(import \"debug\" \"printMemHex\" (func $printMem (param i32 i32)))");
+        imports.push_back("(import \"debug\" \"print\" (func $print (param i32)))");
+        imports.push_back(
+            "(import \"debug\" \"evmTrace\" (func $stackTrace (param i32 i32 i32 i32)))");
+    }
+    imports.push_back("(import \"ethereum\" \"useGas\" (func $useGas (param i64)))");
+
+    wastStr = wast.str();
+    funcs.push_back(wastStr);
+    wastStr = buildModule(funcs, imports, callbackTable);
+    return wastStr;
 }
 
+// given an array for segments builds a wasm module from those segments
+// @param {Array} segments
+// @return {String}
+std::string assembleSegments(const std::vector<JumpSegment>& segments)
+{
+    auto wasm = buildJumpMap(segments);
+
+    for (size_t index; index < segments.size(); ++index)
+    {
+        wasm = "(block ${index} {wasm}"_format("index"_a = index + 1, "wasm"_a = wasm);
+    }
+
+    std::string result =
+        "\
+  (func $main\
+    (export \"main\")\
+    (local $jump_dest i32)\
+    (set_local $jump_dest (i32.const -1))\
+\
+    (block $done\
+      (loop $loop\
+        {wasm}"_format("wasm"_a = wasm);
+    return result;
+}
+
+string evm2wasm(const string& input)
+{
+    return wast2wasm(evm2wast(input));
+}
+
+std::string opcodeToString(opcodeEnum opcode)
+{
+    switch (opcode)
```

Comment:
How about a map for this?

Reply:
I wanted to avoid callers of this function to deal with `std::pair`, but I can change it to a map if you prefer.